### ISSUE TITLE
Jackett memory leak

### ISSUE THREAD

**Issue by dolohow** [Jackett/Jackett#12484]
### Have you checked our Troubleshooting page for your issue?

- [X] I have checked the Troubleshooting page

### Is there already an issue for your problem?

- [X] I have checked older issues, open and closed

### Have you read our Contributing Guidelines?

- [X] I have read the Contributing Guidelines

### Environment

```markdown
Using HTTP Client: HttpWebClient2
Using FlareSolverr: http://localhost:8191/v1
Using proxy: Disabled
App config/log directory: /config
ThreadPool MaxThreads: 32767 workerThreads, 1000 completionPortThreads
Running in Docker: No
File /etc/issue: Ubuntu 20.04.3 LTS \n \l
Jackett variant: CoreLinuxAmdx64
OS version: Unix 5.11.0.38 (64bit OS) (64bit process)
Environment version: 5.0.11 (/app/)
Jackett Data will be stored in: /config
Starting Jackett v0.19.96
```
I actually run in podman instead of docker, but its mostly the same thing

### Description

Jackett leaks memory when doing search over all indexers.  It also happens when I click test all indexers in radarr.

Problems first occurred around 25th October. I update daily

It takes around 9 GB of memory when doing search. After operation is completed, the memory is not released.

EDIT
I narrowed it down, it only happens when I test in radarr one specific indexer - polishsource.  Others seem not to trigger such problem.

EDIT 2
I also quickly bisected and it looks like problem was introduced in version 0.19.48
0.19.34 GOOD
0.19.48 BAD

### Logged Error Messages
```
System.OutOfMemoryException: Insufficient memory to continue the execution of the program.
   at System.Text.StringBuilder.ExpandByABlock(Int32 minBlockCharCount)
   at System.Text.StringBuilder.Append(Char* value, Int32 valueCount)
   at System.Text.StringBuilder.Append(Char[] value, Int32 startIndex, Int32 charCount)
   at System.IO.StringWriter.Write(Char[] buffer, Int32 index, Int32 count)
   at System.Xml.XmlEncodedRawTextWriter.FlushBuffer()
   at System.Xml.XmlEncodedRawTextWriter.WriteElementTextBlock(Char* pSrc, Char* pSrcEnd)
   at System.Xml.XmlEncodedRawTextWriter.WriteString(String text)
   at System.Xml.XmlWellFormedWriter.WriteString(String text)
   at System.Xml.Linq.ElementWriter.WriteElement(XElement e)
   at System.Xml.Linq.XElement.WriteTo(XmlWriter writer)
   at System.Xml.Linq.XContainer.WriteContentTo(XmlWriter writer)
   at System.Xml.Linq.XNode.GetXmlString(SaveOptions o)
   at System.Xml.Linq.XNode.ToString()
   at Jackett.Common.Models.ResultPage.ToXml(Uri selfAtom) in /home/vsts/work/1/s/src/Jackett.Common/Models/ResultPage.cs:line 117
   at Jackett.Server.Controllers.ResultsController.Torznab(TorznabRequest request) in /home/vsts/work/1/s/src/Jackett.Server/Controllers/ResultsController.cs:line 422
```

**Comment by a-nych** [comment]
I'm heaving the same issue with PolishSource, Radarr timesouts because of this.

No other indexers are being used.

**Comment by ilike2burnthing** [comment]
@a-nych can you please provide the details of the system Jackett is running on? The best way to do this is to restart Jackett, access the dashboard, view the log, and copy/paste the last 10 Info lines at the bottom of the log (see the the Environment section of the OP above).

Can you also confirm that rolling back to v0.19.34 resolves this?

**Comment by a-nych** [comment]
<details>
<summary> Log File </summary>
```
10-31 07:02:35 Info Starting Jackett v0.19.34
10-31 07:02:37 Info Environment version: 5.0.11 (/app/Jackett/)
10-31 07:02:37 Info OS version: Unix 5.10.28.0 (64bit OS) (64bit process)
10-31 07:02:37 Info Jackett variant: CoreLinuxAmdx64
10-31 07:02:37 Info File /etc/issue: Ubuntu 18.04.6 LTS \n \l
10-31 07:02:37 Info Running in Docker: Yes (image build: v0.19.34-ls48)
10-31 07:02:37 Info ThreadPool MaxThreads: 32767 workerThreads, 1000 completionPortThreads
10-31 07:02:37 Info App config/log directory: /config/Jackett
10-31 07:02:37 Info Using proxy: Disabled
10-31 07:02:37 Info Using FlareSolverr: No
10-31 07:02:37 Info Using HTTP Client: HttpWebClient2
10-31 07:02:37 Info Loading Native indexers ...
10-31 07:02:37 Info Loaded 96 Native indexers: abnormal, alpharatio, anidex, anidub, AniLibria, animebytes, Animedia, animetorrents, anthelion, avistaz, bakabt, bb, beyond-hd-api, bitcityreloaded, bithdtv, bit-titan, bjshare, broadcasthenet, brokenstones, cgpeers, cinecalidad, cinemaz, danishbytes, darmowetorenty, digitalcore, divxtotal, epublibre, erai-raws, exoticaz, filelist, funfile, fuzer, gazellegames, gimmepeers, hdbitsapi, hdspace, hdtorrents, hebits, immortalseed, internetarchive, iptorrents, libble, lostfilm, magnetico, mejortorrent, morethantv, myanonamouse, ncore, nebulance, newpct, norbits, oppaitime, orpheus, passthepopcorn, piratethenet, pixelhd, polishtracker, pornolab, pretome, privatehd, rarbg, redacted, retroflix, revolutiontt, rutracker, scenehd, scenetime, seals, secretcinema, sharewoodapi, shazbat, ShizaProject, showrss, solidtorrents, speedapp, speedcd, subsplease, superbits, toloka, torrentbytes, torrentday, torrentech, torrentheaven, torrentleech, torrentmafya, torrentnetwork, torrent-paradise-ml, torrentscsv, torrentseeds, torrentsyndikat, tvstore, tvvault, xspeeds, xthor-api, zonaq, animetosho
10-31 07:02:37 Info Loading Cardigann indexers from: /config/cardigann/definitions/, /etc/xdg/cardigan/definitions/, /app/Jackett/Definitions
10-31 07:02:41 Info Loaded 468 Cardigann indexers: 01torrent, 0daykiev, 0magnet, 1337x, 1ptbar, 2fast4you, 3changtrai, 3dtorrents, 4thd, 52pt, 7torrents, Bittorrentfiles, abtorrents, acgrip, acgsou, acidlounge, acrossthetasman, aftershock, aidoruonline, aither-api, aither, amigosshare, anaschcc, anime-free, animeclipse, animeworld-api, animeworld, aniRena, anisource, arabafenice, arabp2p, arenabg, asiancinema, asiandvdclub, audiobookbay, audionews, aussierules, backups, badasstorrents, baibako, bangumi-moe, beitai, beyond-hd-oneurl, beyond-hd, bibliotik, bigfangroup, bitded, bithorlo, bithumen, bitru, bitsearch, bitsexy, bitspyder, bitturk, bluebirdhd, blutopia-api, blutopia, booktracker, bootytape, borgzelle, boxingtorrents, brasiltracker, broadcity, brsociety-api, brsociety, bt4g, btdigg, btetree, btnext, btschool, btsow, bwtorrents, byrutor, carpathians, carphunter, cartoonchaos, casatorrent, cathoderaytube, catorrent, ccfbits, ceskeforum, chdbits, chilebt, cinemageddon, cinemamovies, cinematik, classix, coastalcrew, comicat, concen, concertos, cpasbien, cpasbienclone, crazyhd, crazyspirits, crnaberza, crt2fa, cryptichaven, dariustracker, dark-shadow, darktracker, datascene-api, datascene, datatalli, deildu, demonoid, desireleasers, desitorrents-api, desitorrents, devilsplayground, diablotorrent, dimeadozen, discfan, divteam, dmhy, dragonworldreloaded, dxp, ebookparadijs, ebooks-shares, efectodoppler, ehentai, elitetorrent-biz, empornium, empornium2fa, eniahd, esharenet, estone, ettv, exkinoray, extratorrent-cd, extratorrent-st, extremebits, extremetorrents, extremlymtorrents, exttorrents, eztv, falkonvisionteam, fanoin, fantasticheaven, fantastiko, femdomcult, filebase, filelisting, film-paleis, finelite, finvip, firebit, fouducinema, frozenlayer, funkytorrents, gamestorrents, gay-torrents, gay-torrentsorg, gaytorrentru, generationfree, genesismovement, gfxpeers, gigatorrents, girotorrent, gktorrent, glodls, greekdiamond, greekteam, gtorrentpro, haidan, haitang, hamsterstudio, happyfappy, hd4fans, hdai, hdarea, hdatmos, hdbits, hdbitscom, hdc, hdcenter, hdchina, hdcity, hdcztorrent, hddolby, hdforever, hdhome, hdhouse, hdme, hdolimpo, hdonly, hdroute, hdsky, hdspain, hdtime, hdtorrentsit, hdturk, hdu, hdzone, hellastz, hqsource, huntorrent, ibit, idope, ilcorsaroblu, ilcorsaronero, immortuos, insanetracker, isohunt2, itorrent, iv-torrents, jpopsuki, jptv, karagarga, keepfriends, kickasstorrents-to, kickasstorrents-ws, kinonavse100, kinorun, kinozal, lastfiles, lat-team, learnbits, learnflakes, legittorrents, lemonhd, lepornoinfo, lesaloon, libranet, limetorrents, linkomanija, linuxtracker, losslessclub, mactorrentsdownload, madsrevolution, magicheaven, magico, majomparade, marinetracker, megapeer, mesevilag, metaltracker, microbit, mikan, milkie, mircrew, mixtapetorrent, mma-torrents, mnv, mononokebt, mousebits, moviesdvdr, movietorrent, mteamtp, mteamtp2fa, muziekfabriek, mvgroupforum, mvgroupmain, mypornclub, myspleen, nbtorrents, netcosmo, nethd, netlab, newpct-me, newretro, newstudio, newstudiol, nipponsei, nitro, nntt, noname-club, noname-clubl, nyaasi, oasis, omgwtftrackr, oncesearch, onejav, opencd, oshenpt, ourbits, oxtorrent, p2pbg, p2pelite, parnuxi, pctorrent, peerjunkies, peersfm, pier720, piratadigital, piratbit, piratbitl, pirateiro, pixelcove, polishsource, pornbay, pornbits, pornforall, pornleech, pornolive, pornorip, pornoslab, pornotor, pornotorrent, portugas, potuk, proaudiotorrents, proporno, prostylex, pterclub, ptfiles, pthome, ptmsg, ptsbao, pttime, puntotorrent, purovicio, pussytorrents, puurhollands, pwtorrents, r3vwtf, racing4everyone, racingforme, rainbowtracker, rapidzona, redbits, redstartorrent, reelflix-api, reelflix, renegade, resurrectthenet, rgfootball, rintor, rintornet, riperam, RockBox, romanianmetaltorrents, rptorrents, rus-media, rustorka, rutor, rutracker-ru, scenepalace, scenerush, sdbits, seedfile, selezen, sexypics, shareisland-api, shareisland, sharewood, shokweb, siambit, sktorrent-org, sktorrent, slosoul, snowpt, sosulki, soulvoice, soundpark, speedmasterhd, speedtorrentreloaded, spiritofrevolution, sporthd, sportscult, springsunday, sugoimusic, sukebeinyaasi, swarmazon, tapochek, tasmanit, teamhd, teamos, tekno3d, tellytorrent-api, tellytorrent, teracod, theaudioscene, theempire, thefallingangels, thegeeks, thehorrorcharnel, theleachzone, themixingbowl, theoccult, thepiratebay, theplace, thesceneplace, theshinning, theshow, thevault, tjupt, tlfbits, tokyotosho, torlock, toros, torrent-explosiv, torrent-pirat, torrent-turk, torrent9, torrent9clone, torrentbd, torrentccf, torrentdb, torrentdownload, torrentdownloads, torrentfunk, torrentgalaxy, torrenthr, torrenting, torrentkitty, torrentland, torrentleech-pl, torrentlt, torrentmax, torrentoyunindir, torrentparadise, torrentproject2, torrentqq, torrentsectorcrew, torrentsir, torrentslocal, torrentsurf, torrentv, torrentview, torrentwhiz, torrentz2eu, totallykids, totheglory, totheglorycookie, trackeros, trancetraffic, trezzor, tribalmixes, trupornolabs, ttsweb, turkseed, turktorrent, tvchaosuk, twilight, twilightszoom, u2, uhdbits, underverse, underversel, uniondht, unionfansub, uniongang, unlimitz, vizuk, vsthouse, vsttorrents, wdt, wihd, woot, xiteme, xbytesv2, xwtclassics, xwtorrents, xxxadulttorrent, xxxtor, xxxtorrents, ydypt, yggcookie, yggtorrent, yourbittorrent, yts, zamundanet, zelkaorg, zetorrents, zomb, zooqle, ztracker
10-31 07:02:41 Info Loaded 564 indexers in total
10-31 07:02:41 Info Adding aggregate indexer ('all' indexer) ...
10-31 07:02:41 Info Adding filter indexer ('type:public' indexer) ...
10-31 07:02:41 Info Adding filter indexer ('type:private' indexer) ...
10-31 07:02:41 Info Adding filter indexer ('type:semi-public' indexer) ...
10-31 07:02:41 Info Jackett startup finished in 8.017 s
Hosting environment: Production
Content root path: /app/Jackett/Content
Now listening on: http://[::]:9117
Application started. Press Ctrl+C to shut down.
```
</details>
Yes, rolling back to `v0.19.34` fixed the issue.

**Comment by ilike2burnthing** [comment]
If you haven't already, volume map the container folder `/config` to somewhere accessible on your host, e.g. `/volume1/Docker/Jackett/config`.

Access `/volume1/Docker/Jackett/config` (or wherever you've set) and create a new folder in it called `cardigann` then a new folder within that one called `definitions` (so you end up with `/volume1/Docker/Jackett/config/cardigann/definitions`).

Download https://github.com/Jackett/Jackett/raw/master/src/Jackett.Common/Definitions/polishsource.yml (you can right-click and 'save link as') and rename it to `polishsource48.yml`. Open it in a text editor and change the name and id to `polishsource48`.

Do the same with https://github.com/Jackett/Jackett/raw/72be4c3b0062c5668461007f43a5796b928db267/src/Jackett.Common/Definitions/polishsource.yml but using `polishsource34` instead.

Copy those files to your newly created `/volume1/Docker/Jackett/config/cardigann/definitions` and restart the Jackett container.

Using Jackett v0.19.34 add polishsource48, then add this to Radarr as well and disable the normal polishsource. Run a search through Radarr and see if this causes the same issue. Repeat this with Jackett v0.19.48 and polishsource34. Let me know the results for both.

**Comment by dolohow** [comment]
I did what you wrote me, but there is no `polishsource34` nor `polishsource48` when I try to configure new indexer.

```
docker run -v $HOME/Jackett:/config hotio/jackett:release-0.19.34
```

```
> pwd
$HOME/Jackett
```

```
> ls
cardigann       Indexers  log.txt.20211026.00000.txt  log.txt.20211107.00000.txt  log.txt.20211119.00000.txt
DataProtection  log.txt   log.txt.20211101.00000.txt  log.txt.20211113.00000.txt  ServerConfig.json
```

```
> find cardigann/
cardigann/
cardigann/definitions
cardigann/definitions/polishsource34.yml
cardigann/definitions/polishsource48.yml
```

```
> diff -u cardigann/definitions/polishsource34.yml cardigann/definitions/polishsource48.[17/92986]
--- cardigann/definitions/polishsource34.yml    2021-11-21 14:26:02.052786268 +0100
+++ cardigann/definitions/polishsource48.yml    2021-11-21 14:26:06.016770066 +0100
@@ -1,6 +1,6 @@      
 ---                                                      
-id: polishsource34
-name: PolishSource34
+id: polishsource48                                       
+name: PolishSource48   
 description: "PolishSource (PS) is a POLISH Private Torrent Tracker for MOVIES / TV / GENERAL"
 language: pl-PL
 type: private   
@@ -131,7 +131,7 @@                                                                                                  
     download:
       selector: a[href^="downloadssl.php?id="]
       attribute: href
-    description:                                         
+    genre:                                                                                                          
       optional: true
       selector: td:nth-child(2) > small
       filters:

[output truncated]
```

**Comment by ilike2burnthing** [comment]
And you restarted the container after adding these? Does anything appear in the logs at startup (errors, or them being included in the list of indexers)?

**Comment by dolohow** [comment]
Yes, I restarted and they are not appear in logs, unfortunately.

**Comment by ilike2burnthing** [comment]
> > Also check your env XDG_CONFIG_HOME and XDG_DATA_HOME are set to /config.
> 
> I'm running in container. Is this relevant?

Yes.

**Comment by dolohow** [comment]
I did `docker inspect` and this is what I have:
```
            "Env": [
                "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
                "TERM=xterm",
                "container=podman",
                "CONFIG_DIR=/config",
                "LANGUAGE=en_US:en",
                "XDG_CACHE_HOME=/config/.cache",
                "LC_ALL=en_US.UTF-8",
                "LANG=en_US.UTF-8",
                "APP_DIR=/app",
                "UMASK=002",
                "XDG_DATA_HOME=/config/.local/share",
                "PUID=1000",
                "ARGS=",
                "XDG_CONFIG_HOME=/config/.config",
                "S6_BEHAVIOUR_IF_STAGE2_FAILS=2",
                "BUILD_ARCHITECTURE=linux-amd64",
                "PGID=1000",
                "TZ=Europe/Warsaw",
                "HOME=/root",
                "HOSTNAME=529f05fe99fc"
            ],
```

**Comment by ilike2burnthing** [comment]
There are some there I've not come across before, others are optional, so one or more of these could be causing an issue. To give you something to compare against, mine are:

```none
HOME = /root
LANG = en_US.UTF-8
LANGUAGE = en_US.UTF-8
PATH = /usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
PGID = 999
PUID = 999
TERM = xterm
XDG_CONFIG_HOME = /config
XDG_DATA_HOME = /config
```

Before you go any further, backup your config folder elsewhere.

Try removing all of those variables and just use the defaults here - https://github.com/linuxserver/docker-jackett#usage

If that doesn't work, then add `XDG_CONFIG_HOME` and `XDG_DATA_HOME` with the value `/config`

**Comment by dolohow** [comment]
I guess it could be because I am using _hotio's_ image instead of _linuxerver's_?

**Comment by ilike2burnthing** [comment]
Possibly, feel free to try changing over, but they should be basically identical. The default variables are fairly similar, though the umask value is `002` instead of `022` - https://hotio.dev/containers/jackett/#starting-the-container

**Comment by dolohow** [comment]
> Using Jackett v0.19.34 add polishsource48, then add this to Radarr as well and disable the normal polishsource. Run a search through Radarr and see if this causes the same issue. Repeat this with Jackett v0.19.48 and polishsource34. Let me know the results for both.

`0.19.43` with `polishsource48` does not work
`0.19.48` with `polishsource34` works fine

Since it was a long time ago I did some testing with `0.20.173` and bug is still present

**Comment by ilike2burnthing** [comment]
Can you check if any of the other indexers which were similarly changed in https://github.com/Jackett/Jackett/commit/a01c42c79f87e1bffd06247a69f9d286a415a40c also cause the same issue? Of those, the following are public:
> nipponse, nyaasi, oncesearch, parnuxi, piratbit, pornorip, proporno

This would likely affect any of the commits related to https://github.com/Jackett/Jackett/issues/11917 (1337x to ztracker), but we'll keep it limited to the above for now.

**Comment by dolohow** [comment]
From list above nyaasi works, the rest is not compatible with radarr, so it does not trigger any results, due to being porn or music. 1337x works fine for example

**Comment by ilike2burnthing** [comment]
So this issue only happens when the indexer is used through Radarr, and not when manually searching through Jackett? Is Sonarr also affected?

**Comment by dolohow** [comment]
Yes, issue only happens through Radarr.  Sonarr is not affetected.

**Comment by ilike2burnthing** [comment]
Ok well that makes zero sense to me 😕 

@garfield69 @ngosang @bakerboy448 @Qstick any idea why this would be happening?

TL;DR - memory leak for this user when using polishsource through Radarr and only Radarr, caused by https://github.com/Jackett/Jackett/commit/a01c42c79f87e1bffd06247a69f9d286a415a40c (context - https://github.com/Jackett/Jackett/issues/11917). I can't replicate it, but setups are different.

**Comment by bakerboy448** [comment]
Having the enhanced logs for just a single movie search for the problematic indexer up to and including the OOM crash would probably be helpful no?

and radarr's trace of the search for good measure - i'm assuming radarr 3.2.2.5080 (master & develop currently)?

**Comment by ilike2burnthing** [comment]
@dolohow can you provide the enhanced Jackett and trace Radarr logs requested above?

**Comment by ngosang** [comment]
@ilike2burnthing You can reproduce the issue searching "720p" in PolishSource (Jackett UI).
The issue is there is an infinite loop. In definition field your are using `language and subbed`fields. In those fields your are using `description`. That generates huge strings. Just removing the `description append` from `language and subbed`fields fix the issue.
https://github.com/Jackett/Jackett/blob/a01c42c79f87e1bffd06247a69f9d286a415a40c/src/Jackett.Common/Definitions/polishsource.yml#L159

Was introduced in #11917
You can fix the issue but I would like to handle those loops to avoid future issues. Open a new issue for that if you want.

**Comment by garfield69** [comment]
Yes, I fu**ed this up.
Those appends should not have been carried over when converting the multi-description definition into components assembled into a single description at the end. My bad. I'll fix this.

**Comment by garfield69** [comment]
v0.20.285
